File: internal/storage/types_test.go
package storage

import (
	"testing"
	"time"
)

func TestNodeTypes(t *testing.T) {
	expected := []NodeType{
		"decision",
		"component",
		"convention",
		"concept",
		"session",
		"pattern",
		"issue",
	}

	if len(NodeTypes) != len(expected) {
		t.Errorf("expected %d node types, got %d", len(expected), len(NodeTypes))
	}

	for i, nt := range expected {
		if NodeTypes[i] != nt {
			t.Errorf("expected NodeTypes[%d] = %s, got %s", i, nt, NodeTypes[i])
		}
	}
}

func TestNodeStatuses(t *testing.T) {
	expected := []NodeStatus{"active", "needs_review", "superseded"}

	if len(NodeStatuses) != len(expected) {
		t.Errorf("expected %d statuses, got %d", len(expected), len(NodeStatuses))
	}
}

func TestRelationTypes(t *testing.T) {
	expected := []RelationType{
		"affects",
		"uses",
		"supersedes",
		"relates_to",
		"implements",
		"depends_on",
		"derived_from",
	}

	if len(RelationTypes) != len(expected) {
		t.Errorf("expected %d relation types, got %d", len(expected), len(RelationTypes))
	}
}

func TestKnowledgeNodeFields(t *testing.T) {
	now := time.Now()
	node := KnowledgeNode{
		ID:         "test-id",
		Type:       NodeTypeDecision,
		Title:      "Test Decision",
		Content:    "Test content",
		Tags:       []string{"test"},
		Relations:  []Relation{},
		Confidence: 0.8,
		Created:    now,
		Modified:   now,
		Session:    "session-123",
		Source:     "manual",
		References: []string{"src/test.go"},
		Status:     NodeStatusActive,
	}

	if node.ID != "test-id" {
		t.Errorf("expected ID = test-id, got %s", node.ID)
	}
	if node.Type != NodeTypeDecision {
		t.Errorf("expected Type = decision, got %s", node.Type)
	}
}

func TestCreateKnowledgeNode(t *testing.T) {
	node := CreateKnowledgeNode(
		"test-id",
		NodeTypeDecision,
		"Test Title",
		"Test content",
	)

	if node.ID != "test-id" {
		t.Errorf("expected ID = test-id, got %s", node.ID)
	}
	if node.Type != NodeTypeDecision {
		t.Errorf("expected Type = decision, got %s", node.Type)
	}
	if node.Title != "Test Title" {
		t.Errorf("expected Title = Test Title, got %s", node.Title)
	}
	if node.Confidence != 0.8 {
		t.Errorf("expected Confidence = 0.8, got %f", node.Confidence)
	}
	if node.Source != "manual" {
		t.Errorf("expected Source = manual, got %s", node.Source)
	}
	if node.Status != NodeStatusActive {
		t.Errorf("expected Status = active, got %s", node.Status)
	}
	if len(node.Tags) != 0 {
		t.Errorf("expected empty Tags, got %d", len(node.Tags))
	}
	if len(node.Relations) != 0 {
		t.Errorf("expected empty Relations, got %d", len(node.Relations))
	}
}

func TestUpdateKnowledgeNode(t *testing.T) {
	original := CreateKnowledgeNode("id", NodeTypeDecision, "Title", "Content")

	updated := UpdateKnowledgeNode(original, map[string]any{
		"title": "New Title",
		"tags":  []string{"new-tag"},
	})

	if updated.ID != original.ID {
		t.Errorf("ID should be preserved")
	}
	if updated.Created != original.Created {
		t.Errorf("Created should be preserved")
	}
	if updated.Title != "New Title" {
		t.Errorf("expected Title = New Title, got %s", updated.Title)
	}
	if updated.Modified.Equal(original.Modified) {
		t.Errorf("Modified timestamp should be updated")
	}
}

File: packages/mcp-server/src/storage/types.test.ts
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import {
  NODE_TYPES,
  NODE_STATUSES,
  RELATION_TYPES,
  createKnowledgeNode,
  updateKnowledgeNode,
} from './types.js';

describe('Constants', () => {
  it('should have 7 node types', () => {
    expect(NODE_TYPES).toHaveLength(7);
    expect(NODE_TYPES).toContain('decision');
    expect(NODE_TYPES).toContain('component');
    expect(NODE_TYPES).toContain('convention');
    expect(NODE_TYPES).toContain('concept');
    expect(NODE_TYPES).toContain('pattern');
    expect(NODE_TYPES).toContain('issue');
    expect(NODE_TYPES).toContain('session');
  });

  it('should have 3 node statuses', () => {
    expect(NODE_STATUSES).toHaveLength(3);
    expect(NODE_STATUSES).toContain('active');
    expect(NODE_STATUSES).toContain('needs_review');
    expect(NODE_STATUSES).toContain('superseded');
  });

  it('should have 7 relation types', () => {
    expect(RELATION_TYPES).toHaveLength(7);
    expect(RELATION_TYPES).toContain('affects');
    expect(RELATION_TYPES).toContain('uses');
    expect(RELATION_TYPES).toContain('supersedes');
    expect(RELATION_TYPES).toContain('relates_to');
    expect(RELATION_TYPES).toContain('implements');
    expect(RELATION_TYPES).toContain('depends_on');
    expect(RELATION_TYPES).toContain('derived_from');
  });
});

describe('createKnowledgeNode', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('should create node with required fields', () => {
    vi.setSystemTime(new Date('2024-01-15T10:30:00.000Z'));
    const node = createKnowledgeNode({
      id: 'test-node-12345678',
      type: 'decision',
      title: 'Test Decision',
      content: 'Test content',
    });

    expect(node.id).toBe('test-node-12345678');
    expect(node.type).toBe('decision');
    expect(node.title).toBe('Test Decision');
    expect(node.content).toBe('Test content');
  });

  it('should apply default values', () => {
    vi.setSystemTime(new Date('2024-01-15T10:30:00.000Z'));
    const node = createKnowledgeNode({
      id: 'test-node-12345678',
      type: 'decision',
      title: 'Test',
      content: 'Content',
    });

    expect(node.tags).toEqual([]);
    expect(node.relations).toEqual([]);
    expect(node.confidence).toBe(0.8);
    expect(node.source).toBe('manual');
    expect(node.references).toEqual([]);
    expect(node.status).toBe('active');
  });

  it('should set timestamps to current time', () => {
    vi.setSystemTime(new Date('2024-01-15T10:30:00.000Z'));
    const node = createKnowledgeNode({
      id: 'test-node-12345678',
      type: 'decision',
      title: 'Test',
      content: 'Content',
    });

    expect(node.created).toBe('2024-01-15T10:30:00.000Z');
    expect(node.modified).toBe('2024-01-15T10:30:00.000Z');
  });

  it('should allow overriding defaults', () => {
    vi.setSystemTime(new Date('2024-01-15T10:30:00.000Z'));
    const node = createKnowledgeNode({
      id: 'test-node-12345678',
      type: 'decision',
      title: 'Test',
      content: 'Content',
      tags: ['custom', 'tags'],
      confidence: 0.95,
      source: 'hook_commit',
      status: 'needs_review',
    });

    expect(node.tags).toEqual(['custom', 'tags']);
    expect(node.confidence).toBe(0.95);
    expect(node.source).toBe('hook_commit');
    expect(node.status).toBe('needs_review');
  });
});

describe('updateKnowledgeNode', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('should preserve id and created timestamp', () => {
    vi.setSystemTime(new Date('2024-01-15T10:30:00.000Z'));
    const original = createKnowledgeNode({
      id: 'test-node-12345678',
      type: 'decision',
      title: 'Original',
      content: 'Original content',
    });

    vi.setSystemTime(new Date('2024-01-15T11:00:00.000Z'));
    const updated = updateKnowledgeNode(original, {
      title: 'Updated',
    });

    expect(updated.id).toBe(original.id);
    expect(updated.created).toBe('2024-01-15T10:30:00.000Z');
  });

  it('should update modified timestamp', () => {
    vi.setSystemTime(new Date('2024-01-15T10:30:00.000Z'));
    const original = createKnowledgeNode({
      id: 'test-node-12345678',
      type: 'decision',
      title: 'Original',
      content: 'Original content',
    });

    vi.setSystemTime(new Date('2024-01-15T11:00:00.000Z'));
    const updated = updateKnowledgeNode(original, {
      title: 'Updated',
    });

    expect(updated.modified).toBe('2024-01-15T11:00:00.000Z');
    expect(updated.modified).not.toBe(original.modified);
  });

  it('should update specified fields', () => {
    vi.setSystemTime(new Date('2024-01-15T10:30:00.000Z'));
    const original = createKnowledgeNode({
      id: 'test-node-12345678',
      type: 'decision',
      title: 'Original',
      content: 'Original content',
    });

    const updated = updateKnowledgeNode(original, {
      title: 'Updated Title',
      content: 'Updated content',
      tags: ['new', 'tags'],
    });

    expect(updated.title).toBe('Updated Title');
    expect(updated.content).toBe('Updated content');
    expect(updated.tags).toEqual(['new', 'tags']);
  });

  it('should return new object (immutability)', () => {
    vi.setSystemTime(new Date('2024-01-15T10:30:00.000Z'));
    const original = createKnowledgeNode({
      id: 'test-node-12345678',
      type: 'decision',
      title: 'Original',
      content: 'Original content',
    });

    const updated = updateKnowledgeNode(original, {
      title: 'Updated',
    });

    expect(updated).not.toBe(original);
    expect(original.title).toBe('Original');
    expect(updated.title).toBe('Updated');
  });
});
